### DISCUSSION TITLE
czkawka on android?

### DISCUSSION THREAD

**Discussion by ov3qPI** [qarmin/czkawka#1073] [Ideas]
Would it be feasible to develop an apk of czkawka?

**Comment by qarmin** [comment]
Core functionality - finding duplicates etc. is platform independent so there there is no problem in using it in Android or any other os(supported by Rust).

The problem comes with GUI:
- GTK - which is currently used in GUI not supports android
- I don't have a time to develop gui on this platform(mostly because I don't even know how to do this)

So there is basically 0% chance, that I will do this, but project is open source so anyone can try implement.

**Reply by ov3qPI** [reply]
Very well, thanks for the reply.

So, one could take the core functionality but need to build a new gui for an apk.

**Comment by ov3qPI** [comment]
And for other mobile OS as well.
This has previously been thought of:
https://github.com/qarmin/czkawka/issues/87

**Comment by Quackdoc** [comment]
I tried to run krokiet on android via termux but have run into this issue https://github.com/slint-ui/slint/issues/7145

**Reply by Quackdoc** [reply]
Developing an apk would be a lot of extra work because a lot of applications make a LOT assumptions about android support that make is not really feasible to compile directly to. hoping that using musl or gnu targets on termux will work out.

**Reply by Quackdoc** [reply]
![image](https://github.com/user-attachments/assets/1a5aa706-ca77-4cd2-8c82-59c294dab015)
Got it kind of working, the UI often stops responding, and will always stop working when you add a custom folder path to check. This image is taken over waypipe since it was the only semi-reliable method I have found to use it.

I think slint has some issues when cross compiling for x11 which makes wayland necessary, but wayland is a bit buggy. if anyone else wanted to try it to see if they have better luck

1. Build (no default features is only added for debugging when I was just using software):
`cargo build --release --target aarch64-unknown-linux-gnu --no-default-features --features "winit_software, winit_femtovg"`

2. Send to termux 

3. chmod +x krokiet

4. Install waypipe + termux-glibc + foot for convince (not super necessary but helps with debugging)

5. waypipe:
HOST: /usr/bin/waypipe -s /tmp/socket-local client
HOST: ssh -R /data/data/com.termux/files/usr/tmp/socket-remote:/tmp/socket-local -t -p 8022 u0_a207@192.168.1.180 'XDG_RUNTIME_DIR=$PREFIX/tmp waypipe -s $PREFIX/tmp/socket-remote server -- foot'

6. In terminal:
`XDG_DATA_DIRS=$PREFIX/glibc/share SLINT_DEFAULT_FONT=$XDG_DATA_DIRS/fonts SLINT_BACKEND=winit-femtovg  grun ./krokiet`


Why not musl? When compiling with statically with musl you loose out on dlopen which means you cannot use libwayland or x11, meaning you either need to statically compile them which is not possible ATM. Termux has no musl equivilient to termux-glibc which means dynamically linked is not possible either.

Why not bionic? when you compile for an android target many libraries make assumptions that things like x11 are not available and force you to try and use android surfaces which straight up will not work in termux, and will lead to a long chain of compile time issues. AFAIK rust has no way to override the target_os cfg when you compile using `--target aarch64-linux-android` which I think would actually solve most issues?

I am not sure why x11 isn't working, but I think that might be a lot more stable if I can get it working.

**Comment by Quackdoc** [comment]
I got krokiet working in termux-x11 and termux + ssh. It turns out that winit does not report missing deps properly unless I had run a winit example. How I got czkawka/krokiet working.

```sh
cargo build --release --target aarch64-unknown-linux-gnu
#copy binary to android
ssh -Y -C -p 8022 u0_a207@192.168.1.201  #needs -Y and -C helps a lot over wifi
apt install ttf-dejavu-glibc libxcursor-glibc fontconfig-utils-glibc libxkbcommon-glibc libxi-glibc libx11-glibc grun #it's important to install all of these as slint winit/slint will fail without reporting a proper missing dep
chmod +x ./krokiet
XDG_DATA_DIRS=$PREFIX/glibc/share SLINT_DEFAULT_FONT=$XDG_DATA_DIRS/fonts  RUST_BACKTRACE=1 grun ./krokiet
```


If you plan on using termux-x11 you cannot start krokiet with xstartup (nor would you want to) it is reccomended to use a window manager such as dwm or i3 to make sure you get the full usable space of you phone, then run `XDG_DATA_DIRS=$PREFIX/glibc/share SLINT_DEFAULT_FONT=$XDG_DATA_DIRS/fonts  RUST_BACKTRACE=1 DISPLAY=:1 grun ./krokiet`

I do recommend using SSH if possible since it makes it a lot better experience. Using tethered connection will make this easier, but one of the "remote desktop" solutions may be a lot faster and more stable. 

I haven't managed to get a large amount of images working on a 4G phone, it just OOMs too fast. You will want at least 6G phone and maybe setting

```sh
sudo setprop ro.lmk.use_minfree_levels true # uses old oom method, relevant on A13+
sudo setprop ro.lmk.swap_free_low_percentage 8 #Sets 8% of swap left  before hitting an OOM scenario (Android does NOT like swapping so maybe useless now)
sudo lmkd --reinit #restarts lmkd with new settings
```

EDIT: It may be better to just use sshfs via termux

**Reply by restartxx** [reply]
Would you mind sharing your build/binaries? Please

**Reply by Quackdoc** [reply]
keep in mind this is only going to be useful on arm64 devices but here it is, curl or wget it into termux, chmod +x and excute as I stated how to https://files.catbox.moe/8euua7

**Reply by restartxx** [reply]
> keep in mind this is only going to be useful on arm64 devices but here it is, curl or wget it into termux, chmod +x and excute as I stated how to https://files.catbox.moe/8euua7

Thank you so much for sharing! And sorry for the delay in thanking you lol  😅
I also managed to build one for myself too, the cli and the guis, on my proot-distro and run it from there.

**Reply by robertkirkman** [reply]
Hi, czkawka has a repology score of 20 and to run on android requires patches (like you mentioned), this means that it's a viable candidate for inclusion in the `pkg install` command of Termux, if any of you use czkawka or krokiet on Android frequently and would prefer if `pkg install czkawka` or `pkg install krokiet` worked in Termux, let me know

**Reply by robertkirkman** [reply]
I noticed you used `pkg install glibc-repo` followed by `--target aarch64-unknown-linux-gnu`, that's understandable but it's ok because with enough work, I've been able to compile and package other Rust X11 programs that use Winit or GTK while using `--target aarch64-linux-android`, so if there are a lot of people who want czkawka or krokiet on Termux then I might be able to help with that

**Reply by Quackdoc** [reply]
That would be amazing
